x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsSource.java

```diff
@@ -55,29 +68,46 @@ public DataFrameAnalyticsSource(String[] index, @Nullable QueryProvider queryPro
             throw new IllegalArgumentException("source.index must contain non-null and non-empty strings");
         }
         this.queryProvider = queryProvider == null ? QueryProvider.defaultQuery() : queryProvider;
+        if (sourceFiltering != null && sourceFiltering.fetchSource() == false) {
+            throw new IllegalArgumentException("source._source cannot be disabled");
+        }
+        this.sourceFiltering = sourceFiltering;
     }
 
     public DataFrameAnalyticsSource(StreamInput in) throws IOException {
         index = in.readStringArray();
         queryProvider = QueryProvider.fromStream(in);
+        if (in.getVersion().onOrAfter(Version.CURRENT)) {
+            sourceFiltering = in.readOptionalWriteable(FetchSourceContext::new);
+        } else {
+            sourceFiltering = null;
+        }
     }
 
     public DataFrameAnalyticsSource(DataFrameAnalyticsSource other) {
         this.index = Arrays.copyOf(other.index, other.index.length);
         this.queryProvider = new QueryProvider(other.queryProvider);
+        this.sourceFiltering = other.sourceFiltering == null ? null : new FetchSourceContext(
+            other.sourceFiltering.fetchSource(), other.sourceFiltering.includes(), other.sourceFiltering.excludes());
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeStringArray(index);
         queryProvider.writeTo(out);
+        if (out.getVersion().onOrAfter(Version.CURRENT)) {
+            out.writeOptionalWriteable(sourceFiltering);
+        }
     }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
         builder.array(INDEX.getPreferredName(), index);
         builder.field(QUERY.getPreferredName(), queryProvider.getQuery());
+        if (sourceFiltering != null) {
+            builder.field(_SOURCE.getPreferredName(), sourceFiltering);
```

Comment:
Is there an equivalent of `match_all` source filtering we could report here? Just like we do we `query` (even if user didn't provide one)?

Reply:
There is but its `toXContent()` method is quite verbose:

```
{
  "includes": [],
  "excludes": []
}
```

I assume this is the reason we also avoid to show `analyzed_fields` when it's not set.

Reply:
Ok, let's keep it as it is then.